### PR TITLE
[Feature] Support array functions for JSON type

### LINKED ISSUE #13153: Array functions for JSON type
## Feature request

Several array functions should support `JSON` type:
- `array_agg`
- `array_sort`
- `array_concat`
- `array_slice`
- `reverse`

A typical usage of these functions could be aggregating fields into a `ARRAY<JSON>` and do some calculation on it. 

```sql
-- create table
 CREATE TABLE `t4` (
  `id` int(11) NULL ,
  `t1` int(11) NULL ,
  `t2` int(11) NULL
) ENGINE=OLAP
DUPLICATE KEY(`id`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`id`) BUCKETS 1;

-- load data
INSERT INTO t4 VALUES 
(1 , 1 , 1 ),
(1 , 2 , 2 ), 
(1 , 3 , 3 ),
(2  ,3 , 3 ),
(2 , 1 , 2 ),
(2 , 4 , 1 );

-- Aggregate into Array<JSON>
MySQL > select id, array_agg(json_object('t1', t1, 't2', t2)) from t4 group by id;
+----+------------------------------------------------------------------+
| id | array_agg(json_object('t1', t1, 't2', t2))                       |
+----+------------------------------------------------------------------+
| 2  | ['{"t1": 3, "t2": 3}','{"t1": 1, "t2": 2}','{"t1": 4, "t2": 1}'] |
| 1  | ['{"t1": 1, "t2": 1}','{"t1": 2, "t2": 2}','{"t1": 3, "t2": 3}'] |
+----+------------------------------------------------------------------+
2 rows in set

-- Sort the Array<JSON>
MySQL root@127.1:hehe> select id, array_sort(array_agg(json_object('t1', t1, 't2', t2))) from t4 group by id;
+----+------------------------------------------------------------------+
| id | array_sort(array_agg(json_object('t1', t1, 't2', t2)))           |
+----+------------------------------------------------------------------+
| 2  | ['{"t1": 1, "t2": 2}','{"t1": 3, "t2": 3}','{"t1": 4, "t2": 1}'] |
| 1  | ['{"t1": 1, "t2": 1}','{"t1": 2, "t2": 2}','{"t1": 3, "t2": 3}'] |
+----+------------------------------------------------------------------+
2 rows in set

-- Reverse/Concat/Transform the Array<JSON>


```

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #13153

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 3 / 3 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 7 (57.14%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|4|25.00%|[725, 726, 728]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -720,6 +720,12 @@ public Void visitFunctionCall(FunctionCallExpr node, Scope scope) {
                             fnName + " requires second parameter must be greater than 0");
                 }
                 fn = Expr.getBuiltinFunction(fnName, argumentTypes, Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF);
+            } else if (fnName.equals(FunctionSet.ARRAY_SLICE)) {
+                // Default type is TINYINT, it would match to a wrong function
+                for (int i = 1; i < argumentTypes.length; i++) {
+                    argumentTypes[i] = Type.BIGINT;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:726]: Usually a constant, such as `array_slice(col, 3, 2)`

[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:726]: And the constant is usually `TINYINT`, which would introduce a function matching issue. So here I choose to cast it to a bigint

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 7 (57.14%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|1|4|25.00%|[725, 726, 728]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

**Review by satanson — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**